### ISSUE TITLE
Unseen symbols during inference

### ISSUE THREAD

**Issue by kylebgorman** [CUNY-CL/yoyodyne#7]
[copied from CUNY-CL/abstractness/issues/134]

Currently, the presence of a symbolsnot seen during training causes crash during transducer inference, and I am not sure what happens in the same situation in other architectures. I can imagine two solutions:

1. One could provide an "unk"-ing script which takes train, dev, and test, and then "unk"s (i.e., replaces with the reserved `<UNK>` symbol) any symbols present in dev or test but not in train, and also any symbol below a certain threshold (in case one wants to learn embeddings for this unk character).
2. `yoyodyne-predict` could insert `<UNK>` programmatically where needed during inference.

I think I prefer (2), it's one less step for the user and the unk embedding thing is a bit exotic for our small-vocabulary setting.

**This is a blocker for a post-beta release candidate.**

**Comment by bonham79** [comment]
It would cause a crash in any architecture that has separate features (i.e. pointer-generator). Will be ignored by those models that don't distinguish. However, still a bug since the "<UNK>" character is for input symbols, not features. Likely not the best idea. 

Recent PR makes it that any new feature in validation/test raises an error in the dataloader.

**Comment by Adamits** [comment]
@bonham79 I think one thing that is unique in the transducer though is that it builds an action vocabulary. I believe the action vocabulary will never have an UNK in it, whereas the ptr-generator has *only* the dataset vocabulary, which has an UNK inserted into it manually when creating the `Dataset` class.

I think the ptr-gen issue is related specifically to an UNK feature (as opposed to an UNK character) right?

**Comment by kylebgorman** [comment]
The current behavior (crash with an informative error message on separate-feature models) is superior to the previous state of affairs, but I think it should be handled programmatically and consistently across models instead of resulting in an error.

**Comment by bonham79** [comment]
@Adamits correct, the action vocabulary only will produce vocabulary seen in the training set and has no equivalent of <UNK> (outside of simply copying or deleting an UNK character). 

The ptr-gen issue would be specifically the UNK feature issue. While other models have a more subtler issue of using the UNK _character_ embedding for an UNK feature. This is currently not an issue since the raise error flag stops passing of new features.

To avoid the crash, we just need to create a new symbol in SPECIAL for UNK_FEAT and then use that when building the feature vocabulary. This would need to be followed by a few experiments that compare use of UNK_FEAT against simply not having that feature present in validation. This is because I have an intuition that the embedding won't get much training time and so won't really add much to accuracy. (May actually hurt.)

Alternatively, we could build into base model architecture a function that averages all features for an UNK_FEAT embedding. This would create a scenario where the UNK_FEAT gets (technically) more exposure during training and thus be more useful for validation.

**Comment by bonham79** [comment]
> Alternatively, we could build into base model architecture a function that averages all features for an UNK_FEAT embedding. This would create a scenario where the UNK_FEAT gets (technically) more exposure during training and thus be more useful for validation.

adding onto this, i'm wondering if we shouldn't just make this also the default behavior for UNK characters as well? as we have it, I don't think an UNK embedding has any means of training so maintains randomness.

**Comment by kylebgorman** [comment]
My bias is to keep things simple as possible here: my only interest is that I want the system to run, for all architectures, when there are unseen symbols present.  

I don't take it as obvious that unseens should be expected to behave like the mean symbol. (Indeed it's very possibly a symbol out of left field.) I suspect that any way we make an unseen embedding, it'll be uninformative and undertrained, and that any input or output containing an unseen will be mostly mispredicted no matter what we do. So I'd be very happy with one, or many, random embeddings for now.

Of course I could be wrong but I'd prefer a simple solution now (as I consider this a serious blocker) and we can try fancier things later once we're convinced they improve things.

Another minor issue for UNK target symbols: no system should be able to get credit or predicting UNK. That is, in accuracy computation, UNK != UNK, like NaN.

**Comment by bonham79** [comment]
Closing issue with https://github.com/CUNY-CL/yoyodyne/pull/18
